### ISSUE TITLE
podman load: unexpected repository path

### ISSUE THREAD

**Issue by edsantiago** [projectatomic/libpod#667]
Attached is a very trivial image created by `docker save`. After loading it, podman reports it having an unexpected repo path:
```
# podman load --input /tmp/load_test.tar.gz
Getting image source signatures
Copying blob sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef
 1024 B / 1024 B [==========================================================] 0s
Skipping fetch of repeat blob sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef
Skipping fetch of repeat blob sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef
Copying config sha256:5a989dd2c48f406db792ebb7fca50160a67398db0af6f15c806f609b77021e9f
 1.58 KB / 1.58 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
Loaded image:  load:test
```
So far, so good: `load:test` is the expected tag. But:
```
# podman images
REPOSITORY               TAG    IMAGE ID       CREATED       SIZE
docker.io/library/load   test   5a989dd2c48f   3 years ago   5.44kB
^^^^^^^^^^^^^^^^^^  <----
```
Whence the `docker.io/library/` ? docker output for reference:
```
# docker images
REPOSITORY              TAG   IMAGE ID      CREATED       SIZE
load                    test  5a989dd2c48f  3 years ago   0 B
```

[load_test.tar.gz](https://github.com/projectatomic/libpod/files/1944502/load_test.tar.gz)

**Comment by mheon** [comment]
Related: #639

**Comment by edsantiago** [comment]
Could someone with privs please reopen this, and remove the `duplicate` tag?

I'm catching up on backlog, and this issue is still present in podman-0.7.1-1.git802d4f2.fc28. It turns out not to be quite a dup of #639 after all.

**Comment by mheon** [comment]
@rhatdan Looks like we didn't quite manage to get all the `docker.io` references out

**Comment by rhatdan** [comment]
@haircommander looks like podman load still hard codes docker.io/library onto loaded images. Could you take a quick look at this?
@umohnani8 Will probably be able to help.
After this load the value should be localhost/load

**Comment by haircommander** [comment]
The problem is in github.com/containers/image/docker/reference/normalize.go:splitDockerDomain
When we load a new image, it goes through this stack trace:
cmd/podman/load.go:108 -> runtime.ImageRuntime().LoadFromArchive(...)
libpod/image/image.go:182 -> newImage.getLocalImage()
libpod/image/image.go:213 -> i.imageruntime.getImage(...)
libpod/image/image.go:372 -> is.Transport.ParseStoreReference(...)
vendor/github.com/containers/image/storage/storage_transport.go:166 -> reference.ParseNormalizedNamed()
vendor/github.com/containers/image/docker/reference/normalize.go:36 -> splitDockerDomain()
vendor/github.com/containers/image/docker/reference/normalize.go:70 checks for localhost prepended

Since at no place in this process have we prepended local host to it, it prepends "defaultDomain" which in this case is "docker.io/library".

This behavior was also found using buildah bud, so it's clearly entrenched. I am not sure the best way to proceed.

**Comment by haircommander** [comment]
One potential fix is changing the vendored code at containers/image/storage/storage_transport.go:163 adding:
```
        domainDoesNotExist := (strings.IndexRune(ref, '/') < 1)
        if domainDoesNotExist {
            ref = "localhost/" + ref
        }
```
But it breaks the images tests so I don't know if that's the route we want. I can revisit it tomorrow if needed.

**Comment by rhatdan** [comment]
@mtrmac WDYT?

**Comment by rhatdan** [comment]
Can we force the localhost prepend?

**Comment by haircommander** [comment]
@baude I am not clear on what you asked

**Comment by mtrmac** [comment]
**NOTE: All of this is assuming that the `localhost` prefixing is desirable**.

(For reference, earlier discussions in https://github.com/containers/image/pull/448 https://github.com/projectatomic/libpod/issues/639 , https://github.com/projectatomic/buildah/pull/648 , and elsewhere. I can’t remember _all_ of it, assuming that these are the latest and reflect what we learned before.)

I would, in principle, not be opposed to `storageReference` parsing using a `localhost` default/normalization instead of a `docker.io` one. At least going by the references above, I can’t remember precisely why that didn’t happen (my best guess is that it was insufficient to make search by short names work, and once that was implemented, normalization at the `storageReference` level seemed unnecessary; I could well be wrong).

As far as the implementation goes, though, I do think that such a mapping, if any, needs to happen at a consistent layer of the code, and as of https://github.com/projectatomic/buildah/pull/648/files , which happened instead, the current `buildah` implementation ties the `localhost` default fairly tightly to the search behavior , and if I understandmakes the normalization user-choosable via `imagebuildah.BuildOptions.Registry`. `c/image/storage` can’t very well normalize based on `imagebuildah.BuildOptions`, and because a normalization in that package is exposed in a CLI (though `alltransports.parseImageName`), I’d prefer it not to be variable at that level of the stack.

---

(
```
domainDoesNotExist := (strings.IndexRune(ref, '/') < 1) …
```

This would not be correct; `redhat/rhel` is a valid hostname-unqualified name. Compare the conditions in `c/image/docker/reference.splitDockerDomain`.
)

**Comment by mtrmac** [comment]
> Attached is a very trivial image created by docker save.

… which contains `repositories`:
```
{"load":{"test":"932…de0"}}
```
and equivalently `manifest.json`:
```
[{…"RepoTags":["load:test"]…}]
```

The thing is, though, that as far as the image format can express, this _is_ `docker.io/library/load:test`.  Upstream `moby/moby` will produce exactly the above data when asked to save `docker.io/library/load:test`, and if you manually modify the tarball to use the fully-qualified format, on `docker load` the two will be treated exactly the same.

`projectatomic/docker`, from reading the code, differs: it will preserve the qualified/unqualified format of the CLI input on `save`, and of the tarball on `load` — but that’s not actually desirable. See the long discussion around https://github.com/containers/image/issues/72#issuecomment-262069191 for details, the short version is that passing unqualified names to `projectatomic/docker load` is a bad idea.

So, as far as both implementation are concerned, unqualified names in `docker save`-formatted tarballs don’t or shouldn’t exist, and treating the value as `docker.io/library/load:test` is the more interoperable and more consistent alternative.

---

Of course, `podman load` is not forced to use exactly the name(s) inside the tarball, or even anything resembling them at all, when forming the destination. (And, really, allowing potentially untrusted input to choose what will be overwritten in users’ destination is a fairly risky design.)  But, if the goal is to interpret the file format faithfully, I think there is a fairly strong case for interpreting it as `docker.io/library/load:test`.

If `podman` wants to interpret it as `localhost/load:test`, to be consistent with things like `buildah commit`, sure, that’s plausible as well.  I think it _should_ be consistent with `buildah commit` etc, though, so interpreting it as `load:test` with no domain does not look so attractive to me.

**Comment by rhatdan** [comment]
Right I want it treated as localhost/load:test

**Comment by baude** [comment]
I think @haircommander is on the right track after our discussion Friday.  @haircommander ... agree or would you like a little help?

**Comment by haircommander** [comment]
Yeah I have the solution, just need to write a unit test and make sure it doesn't break anything.

**Comment by haircommander** [comment]
closed via #1103
